pyrelational/pipeline/generic_pipeline.py

```diff
@@ -0,0 +1,321 @@
+
+    def full_active_learning_run(
+        self,
+        num_annotate: int,
+        num_iterations: Optional[int] = None,
+        oracle_interface: object = None,
+        test_loader: DataLoader = None,
+        return_query_history: bool = False,
+        *strategy_args,
+        **strategy_kwargs,
+    ) -> Optional[Dict]:
+        """Given the number of samples to annotate and a test loader
+        this method will go through the entire active learning process of training
+        the model on the labelled set, and recording the current performance
+        based on this. Then it will proceed to compute uncertainties for the
+        unlabelled observations, rank them, and get the top num_annotate observations
+        labelled to be added to the next iteration's labelled dataset L'. This
+        process repeats until there are no observations left in the unlabelled set.
+
+        :param num_annotate: number of observations to get annotated per iteration
+        :param num_iterations: number of active learning loop to perform
+        :param oracle_interface: undefined for now, this will be entry point for external oracle later
+        :param test_loader: test data with which we evaluate the current state of the model given the labelled set L
+        :param return_query_history: whether to return the history of queries or not
+        :return: optionally returns a dictionary storing the indices of queries at each iteration
+        """
+        iter_count = 0
+        if return_query_history:
+            query_history = {}
+        while len(self.u_indices) > 0:
+            iter_count += 1
+
+            # Obtain samples for labelling and pass to the oracle interface if supplied
+            observations_for_labelling = self.active_learning_step(
+                num_annotate=num_annotate, *strategy_args, **strategy_kwargs
+            )
+            if return_query_history:
+                query_history[iter_count] = observations_for_labelling
+
+            # Record the current performance
+            self.performances[self.iteration] = self.current_performance(
+                test_loader=test_loader,
+                query=observations_for_labelling,
+            )
+
+            self.active_learning_update(
+                observations_for_labelling,
+                oracle_interface,
+                update_tag=str(self.iteration),
+            )
+            if (num_iterations is not None) and iter_count == num_iterations:
+                break
+
+        # Final update the model and check final test performance
+        self.model.train(self.l_loader, self.valid_loader)
+        self.performances[self.iteration] = self.current_performance(test_loader=test_loader)
+        if return_query_history:
+            return query_history
+
+    def update_annotations(self, indices: List[int]) -> None:
+        """
+        Update labels in the datamanager train set based on indices selected for labelling
```

Comment:
Ok I will implement a default oracle which simply passes the `update_dataset` method defined in `oracle.generic_oracle.GenericOracle`, this should take care of the if statement within the generic_pipeline.

As for the first part, I will simply provide a few more details in the docstring. The `update_annotations` method simply calls `data_manager.update_train_labels(indices)` which simply shifts the relevant unlabelled indices to the labelled indices. Provided that the `update_dataset` method is sufficiently implemented there should be no issue with this. To help here I realise it may be useful to have a method that hides having to directly touch the dataset in the data_manager. I can add another method called `update_target(ind, target_value)` which does this bit safely.

This way the required `update_dataset` method for creating will involve just the obtaining label part for whatever process/etc it is to interface and upon obtaining the target values will just call the `update_target(ind, target_value)` method.

Reply:
It may be useful to rename the method as well. Something that makes it clear it's updating data manager based on current run. update_dataset_from_query, or something ?

Reply:
Updated interface in b0d7a61